crates/pgt_completions/src/sanitization.rs

```diff
@@ -0,0 +1,285 @@
+use std::borrow::Cow;
+
+use pgt_text_size::TextSize;
+
+use crate::CompletionParams;
+
+pub(crate) struct SanitizedCompletionParams<'a> {
+    pub position: TextSize,
+    pub text: String,
+    pub schema: &'a pgt_schema_cache::SchemaCache,
+    pub tree: Cow<'a, tree_sitter::Tree>,
+}
+
+impl<'larger, 'smaller> TryFrom<CompletionParams<'larger>> for SanitizedCompletionParams<'smaller>
+where
+    'larger: 'smaller,
+{
+    type Error = String;
+
+    fn try_from(params: CompletionParams<'larger>) -> Result<Self, Self::Error> {
+        let tree = match &params.tree {
+            Some(tree) => tree,
+            None => return Err("Tree required for autocompletions.".to_string()),
+        };
+
+        if cursor_inbetween_nodes(tree, params.position)
+            || cursor_prepared_to_write_token_after_last_node(tree, params.position)
+            || cursor_before_semicolon(tree, params.position)
+        {
+            Ok(SanitizedCompletionParams::with_adjusted_sql(params))
+        } else {
+            Ok(SanitizedCompletionParams::unadjusted(params))
+        }
+    }
+}
+
+static SANITIZED_TOKEN: &str = "REPLACED_TOKEN";
+
+impl<'larger, 'smaller> SanitizedCompletionParams<'smaller>
+where
+    'larger: 'smaller,
+{
+    fn with_adjusted_sql(params: CompletionParams<'larger>) -> Self {
+        let cursor_pos: usize = params.position.into();
+        let mut sql = String::new();
+
+        for (idx, c) in params.text.chars().enumerate() {
+            if idx == cursor_pos {
+                sql.push_str(SANITIZED_TOKEN);
+                sql.push(' ');
+            }
+            sql.push(c);
+        }
+
+        let mut parser = tree_sitter::Parser::new();
```

Comment:
isnt there a way to modify the tree sitter tree instead of reparsing the string?

Reply:
I think editing the node isn to possible, so maybe we should try to pass the existing tree sitter parser down.